### PR TITLE
tools: detector: add support for YDLidar G2

### PR DESCRIPTION
**Pull request by ecourtois**

### CODE REVIEW

File: cogip/tools/lidarusb/lidar.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                    if len(g) == 1:
+                        index = g[0]
+                        isolated = result[index]
+                        before = result[(index - 1) % 360]
+                        after = result[(index + 1) % 360]
+                        result[index] = (
+                            isolated[0],
+                            round((before[1] + after[1]) / 2),
+                            round((before[2] + after[2]) / 2),
+                        )
```
Comment (gdoffe) [inline cogip/tools/lidarusb/lidar.py:113]: is that filter inherited from previous LIDAR code ? Still useful ?

File: cogip/widgets/lidarview.py [hunk 2]
```diff
@@ -23,41 +23,48 @@ class LidarView(QtWidgets.QWidget):
     """
     min_distance: int = 100
     max_distance: int = 5000
-    max_intensity: int = 12000
+    max_intensity: int = 1024
```
Comment (gdoffe) [inline cogip/widgets/lidarview.py:26]: Is it the max intensity you define or the max intensity the LIDAR is capable of ?
maybe need some few small comments here.

File: cogip/tools/detector/detector.py [hunk 3]
```diff
@@ -76,60 +73,82 @@ def __init__(
...
+            self._laser.setlidaropt(ydlidar.LidarPropSerialBaudrate, 230400)
+            self._laser.setlidaropt(ydlidar.LidarPropLidarType, ydlidar.TYPE_TRIANGLE)
+            self._laser.setlidaropt(ydlidar.LidarPropDeviceType, ydlidar.YDLIDAR_TYPE_SERIAL)
+            self._laser.setlidaropt(ydlidar.LidarPropSingleChannel, False)
+            self._laser.setlidaropt(ydlidar.LidarPropSampleRate, 5)
+            self._laser.setlidaropt(ydlidar.LidarPropIntenstiy, True)
+            self._laser.setlidaropt(ydlidar.LidarPropScanFrequency, 5.0)
+            self._laser.setlidaropt(ydlidar.LidarPropMaxRange, self._max_distance/1000)
+            self._laser.setlidaropt(ydlidar.LidarPropMinRange, self._min_distance/1000)
+            self._laser.setlidaropt(ydlidar.LidarPropInverted, False)
```
Comment (gdoffe) [inline cogip/tools/detector/detector.py:101]: Properties seems diffrent here than in `lidar.py`.
Changes are not obvious.

File: cogip/tools/detector/detector.py [hunk 4]
```diff
@@ -179,28 +185,28 @@ def filter_distances(self) -> List[int]:
...
-            dist_min = self._lidar_data[first % self.NB_ANGLES]
+        while first < start + 360:
+            dist_min = self._lidar_data[first % 360]
 
             # Find the next angle with obstacle
             if dist_min >= self._max_distance:
                 first += 1
                 continue
 
             # An angle (first) with obstacle is found, iterate until the next obstacle without obstacle
```
Comment (gdoffe) [inline cogip/tools/detector/detector.py:196]: not in PR but it should be `iterate until the next angle without obstacle` ?

File: cogip/tools/detector/detector.py [hunk 5]
```diff
@@ -267,3 +273,63 @@ def process_lidar_data(self):
...
+        """
+        Start the Lidar.
+        """
+        if self._laser:
+            self._laser.initialize()
+            self._laser.turnOn()
+
+    def read_lidar(self):
+        """
+        Function executed in a thread loop to reader Lidar data.
```
Comment (gdoffe) [inline cogip/tools/detector/detector.py:287]: s/reader/read

File: cogip/tools/detector/detector.py [hunk 6]
```diff
@@ -267,3 +273,63 @@ def process_lidar_data(self):
...
+            for angle, distances in enumerate(tmp_distances):
+                distance = self._max_distance
+                if size := len(distances):
+                    distance = round(sum(distances) * 1000 / size)
+                else:
+                    empty_angles.append(angle)
+
+                result.append(distance + self._beacon_radius)
+
+            # If a degree has not valid and is isolated (no other empty angle before and after)
```
Comment (gdoffe) [inline cogip/tools/detector/detector.py:315]: s/has/is

File: docs/usage/detector.md [hunk 7]
```diff
@@ -13,11 +13,14 @@ and by `mcu-firmware` to compute avoidance path.
 ### Monitoring Mode
 
 Read data from lidar connected on a serial port of the Raspberry Pi.
+This is the default mode when a Lidar is connected, since the YDLidar SDK automatically
```
Comment (gdoffe) [inline docs/usage/detector.md:16]: automatically **select** ?

File: docs/usage/detector.md [hunk 8]
```diff
@@ -13,11 +13,14 @@ and by `mcu-firmware` to compute avoidance path.
...
+the serial port to use.
 
 ### Emulation Mode
 
 Ask the `Monitor` to emulate the Lidar wich sends its data through the SocketIO server.
 The emulation mode is enabled if no Lidar is detected at startup.
+In case a false detection, use the `--emulation` option.
```
Comment (gdoffe) [inline docs/usage/detector.md:23]: In case **of** a false

**Review by gdoffe — APPROVED** [review]
Excellent ! Good to have a real LIDAR. ;)

I approve already but you have few typos and I have some questions.

File: cogip/tools/lidarusb/lidar.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+                    else:
+                        empty_angles.append(angle)
+
+                    result.append((
+                        angle - 180,
+                        distance,
+                        intensity
+                    ))
+
+                # If a degree has not valid and is isolated (no other empty angle before and after)
```
[the remark this answers is not in the document]
Comment (ecourtois) [reply cogip/tools/lidarusb/lidar.py:100]: Fixed: it was `has no valid point`

File: cogip/tools/lidarusb/lidar.py (hunk 1 above)
Comment (ecourtois) [reply cogip/tools/lidarusb/lidar.py:113]: It's a new filter.

File: cogip/widgets/lidarview.py (hunk 2 above)
Comment (ecourtois) [reply cogip/widgets/lidarview.py:26]: Comment added

File: cogip/tools/detector/detector.py (hunk 3 above)
Comment (ecourtois) [reply cogip/tools/detector/detector.py:101]: What is different?

File: cogip/tools/detector/detector.py (hunk 4 above)
Comment (ecourtois) [reply cogip/tools/detector/detector.py:196]: Fixed.

File: cogip/tools/detector/detector.py (hunk 5 above)
Comment (ecourtois) [reply cogip/tools/detector/detector.py:287]: Fixed

File: cogip/tools/detector/detector.py (hunk 6 above)
Comment (ecourtois) [reply cogip/tools/detector/detector.py:315]: Fixed

File: docs/usage/detector.md (hunk 7 above)
Comment (ecourtois) [reply docs/usage/detector.md:16]: Fixed.

File: docs/usage/detector.md (hunk 8 above)
Comment (ecourtois) [reply docs/usage/detector.md:23]: Fixed

File: cogip/widgets/lidarview.py (hunk 2 above)
Comment (ecourtois) [reply cogip/widgets/lidarview.py:26]: Actually it is both: max intensity of the Lidar and max intensity to display.

File: cogip/tools/detector/detector.py (hunk 3 above)
Comment (gdoffe) [reply cogip/tools/detector/detector.py:101]: I dont know, that's the point ! Lets talk about this on discord. ;)

**Pull request merged**